metadata-ingestion/src/datahub/ingestion/source/aws/dynamodb.py

```diff
@@ -0,0 +1,188 @@
+from datahub.metadata.schema_classes import (
+    ChangeTypeClass,
+)
+from datahub.metadata.com.linkedin.pegasus2avro.schema import (
+    BooleanType,
+    BytesType,
+    KeyValueSchema,
+    NumberType,
+    SchemaField,
+    SchemaFieldDataType,
+    SchemaMetadata,
+    StringType,
+)
+
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_PLATFORM = "dynamodb"
+
+_attribute_type_mapping = {
+    "S": StringType,
+    "N": NumberType,
+    "B": BytesType,
+    "BOOL": BooleanType,
+}
+
+
+class DynamoDBSourceConfig(AwsSourceConfig, PlatformSourceConfigBase):
+
+    @property
+    def dynamodb_client(self):
+        return self.get_dynamodb_client()
+
+    @property
+    def dynamodbstreams_client(self):
+        return self.get_dynamodbstreams_client()
+
+
+@dataclass
+class DynamoDBSourceReport(SourceReport):
+    filtered_tables: List[str] = field(default_factory=list)
+    streaming_disabled_tables: List[str] = field(default_factory=list)
+
+    def report_table_filtered(self, table: str) -> None:
+        self.filtered_tables.append(table)
+
+    def report_table_streaming_disabled(self, table: str) -> None:
+        self.streaming_disabled_tables.append(table)
+
+@dataclass
+class DynamoDBSource(Source):
+    config: DynamoDBSourceConfig
+    report: DynamoDBSourceReport()
+
+    def __init__(self, config: DynamoDBSourceConfig, ctx: PipelineContext):
+        super().__init__(ctx)
+        self.config = config
+        self.report = DynamoDBSourceReport()
+
+
+    @classmethod
+    def create(cls, source_config, context):
+        config = DynamoDBSourceConfig.parse_obj(source_config)
+        return cls(config, context)
+
+
+    def get_workunits(self) -> Iterable[MetadataWorkUnit]:
+        for table_name in get_all_tables(self.config.dynamodb_client):
+            logger.info(f"Processing table {table_name}")
+            if not self.config.table_pattern.allowed(table_name):
+                logger.warn(f"Table {table_name} is not allowed by table_allow pattern")
+                self.report.report_table_filtered(table_name)
+                continue
+            table = describe_table(self.config.dynamodb_client, table_name)
+            key_schema = table.get("KeySchema", [])
+            keys = [x["AttributeName"] for x in key_schema]
+            raw_attribute_definitions = table["AttributeDefinitions"]
+            attribute_definitions = {
+                x["AttributeName"]: x["AttributeType"] for x in raw_attribute_definitions
+            }
+            if table.get("StreamSpecification", {}).get("StreamEnabled", False):
+                stream_arn = table["LatestStreamArn"]
+                attribute_definitions.update(
+                    get_attribute_definitions(self.config.dynamodbstreams_client, stream_arn)
+                )
+            else:
+                logger.error(f"Table {table_name} streaming is NOT enabled.")
+                self.report.report_table_streaming_disabled(table_name)
```

Comment:
Do we need a `continue` after this?

Reply:
we can still populate the table schema with the limited `attribute_definitions` of `describe_table` response. but yes, the table schema is not complete.

Reply:
That could be misleading or confusing for someone looking for it on DataHub – as it wouldn't reflect the current state of the world.